designs/0005-cmdstan3.md

```diff
@@ -0,0 +1,161 @@
+---
+output:
+  html_document: default
+  pdf_document: default
+---
+
+- *Feature Name:* CmdStan3
+- *Start Date:* 2020-01-16
+- *RFC PR(S):*
+- *Stan-dev Issue(s):*
+
+# CmdStan3 Summary
+[summary]: #summary
+
+CmdStan3 is a command-line interface to the Stan services layer
+which provides an extensible syntax for calling Stan services
+as well as model compilation.
+CmdStan3 provides a utility program `stan` which takes as arguments
+a set of subcommands; these subcommands map to functions
+exposed by the `stan::services` package, model compilation,
+and existing Stan utilities such as `stansummary`.
+
+
+# Motivation
+[motivation]: #motivation
+
+1. Unify all workflow operations: single utility for model compilation, fitting, and diagnostic operations.
+
+2. Usability: flat command syntax which allows configuration arguments to be specified in any order
+
+3. Extensibility: easy to add new commands corresponding to new services and to modify argument structure for existing commands.
+
+# Guide-level explanation: Command Syntax
+[guide-level-explanation]: #guide-level-explanation
+
+
+The first argument to the `stan` utility is either
+
+- `-h` or `--help` for help
+- `-v` or `--version` for CmdStan3 version information
+- a subcommand
+
+Subcommand arguments are then followed by either
+
+- `-h` or `--help` for subcommand-specific help
+- a Stan model name
+
+For subcommands which call a Stan service,
+the remaining command line arguments consist of an
+unordered set of pairs of the form `<name>=<value>`
+which are parsed and validated by the `stan` utility
+and then passed in to the services method.
+
+Possible exceptions to this subcommand syntax are:
+
+- The `compile` subcommand, which would instead take a series of makefile options.
+- Utilities such as `stansummary` which operate on one or more files but don't need a compiled model.
+
```

Comment:
I am not sure if this is a topic to discuss here but would it be worth it to add a "machine friendly" mode that would print the stdout in such a way that it would be easier to parse in other applications. This would simplify cmdstanX interfaces as well as any other script for running cmdstan others are using.